### PR TITLE
storcon: skip draining shard if it's secondary is lagging too much

### PR DESCRIPTION
**Pull request by VladLazar**
## Problem

Migrations of tenant shards with cold secondaries are holding up drains in during production deployments.

## Summary of changes

If a secondary locations is lagging by more than 256MiB (configurable, but that's the default), then skip
cutting it over to the secondary as part of the node drain.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/secondary/downloader.rs [hunk 1]
```diff
@@ -1146,12 +1146,14 @@ impl<'a> TenantDownloader<'a> {
...
         ctx: &RequestContext,
     ) -> Result<Option<HeatMapLayer>, UpdateError> {
-        // Failpoint for simulating slow remote storage
+        // Failpoints for simulating slow remote storage
         failpoint_support::sleep_millis_async!(
             "secondary-layer-download-sleep",
             &self.secondary_state.cancel
         );
 
+        pausable_failpoint!("secondary-layer-download-pausable");
```
Comment (jcsp) [inline pageserver/src/tenant/secondary/downloader.rs:1155]: Couldn't the existing failpoint above be used?  (the sleep_millis failpoints are usable for pausing if you pass a large value in)

File: storage_controller/src/service.rs [hunk 2]
```diff
@@ -5715,18 +5722,70 @@ impl Service {
...
+                |client| async move { client.tenant_secondary_status(tenant_shard_id).await },
+                &self.config.jwt_token,
+                1,
+                3,
+                Duration::from_millis(250),
+                &self.cancel,
+            )
+            .await
+        {
+            Some(Ok(status)) => Ok(status.bytes_total - status.bytes_downloaded),
```
Comment (jcsp) [inline storage_controller/src/service.rs:5747]: `bytes_total` is only updated when they download a heatmap, so a secondary that has a lot of work to do might not realise they have a lot of work to do until they start trying to download.

So this works as a pre-check for tenants that are definitely a long way from being able to live migrate quickly, but seeing a low 'lag' number here does not guarantee that the live migration will actually complete quickly.

**Comment by jcsp** [comment]
There's a key design decision here: when we can't to a high quality live migration quickly, do we prefer to:
A) leave the attached location where it is
B) to cut it over anyway with a less-than-warm cache

This PR does A: I haven't thought about it deeply but would like to make sure we have a solid rationale.

Whether we prefer A or B, I think we need a way to tell `live_migrate` to give up more quickly than usual if it is doing a migration during a drain-for-restart: today, we have `TOTAL_DOWNLOAD_TIMEOUT` after which we cross out fingers and go ahead to cut over anyway.  We need this because even with the pre-check on pending downloads added in this PR, we can still end up hitting a download that takes a long time (because the pre-check is based on the most recent heatmap the secondary location saw)

TLDR: this PR isn't wrong, and covers the case of shards that are egregiously far behind on downloads (e.g. right after a shard split happens), but I think we need to invest a bit more time to figure out our total story

File: storage_controller/src/service.rs (hunk 2 above)
Comment (jcsp) [reply storage_controller/src/service.rs:5747]: Also: secondary progress can contain zeros for a secondary location that has never done a heatmap download (e.g. a new location, or right after pageserver restart), so need an opinionated handling of that case -- probably assuming it is very laggy makes sense, because we can't trust that it will cut over quickly.

File: pageserver/src/tenant/secondary/downloader.rs (hunk 1 above)
Comment (VladLazar) [reply pageserver/src/tenant/secondary/downloader.rs:1155]: Perhaps. I was unsure if I can turn it off while it's sleeping. I don't want the test to wait on the failpoint.

**Comment by VladLazar** [comment]
> There's a key design decision here: when we can't to a high quality live migration quickly, do we prefer to: A) leave the attached location where it is B) to cut it over anyway with a less-than-warm cache

Indeed. To my mind a hybrid approach is the desired state:
* If the secondary is lagging by a lot (more than 256MiB or 512MiB?) I'd say A is preferable since there's a big chance QOS will suffer more if we migrate. Restarts should be quick if the pageserver is not loaded (i.e. has drained most shards).
* If the secondary is lagging by a solid amount (more than 64MiB), then B is preferable since we can soak up the cost of hydrating on the secondary.
* If the secondary is lagging by a bit (less than 64MiB), then perhaps it makes sense to warm it up before cutting over

This PR implements the first bullet point.
 
> Whether we prefer A or B, I think we need a way to tell `live_migrate` to give up more quickly than usual if it is doing a migration during a drain-for-restart: today, we have `TOTAL_DOWNLOAD_TIMEOUT` after which we cross out fingers and go ahead to cut over anyway. We need this because even with the pre-check on pending downloads added in this PR, we can still end up hitting a download that takes a long time (because the pre-check is based on the most recent heatmap the secondary location saw)

Yes, this makes sense. Will add.

File: storage_controller/src/service.rs (hunk 2 above)
Comment (VladLazar) [reply storage_controller/src/service.rs]: Thanks for flagging. Addressed in https://github.com/neondatabase/neon/pull/8644/commits/69dc6466ce5f17297a2dd646b87418671b520456

**Review by jcsp — APPROVED** [review]
